### PR TITLE
Add support for Apple Swift language

### PR DESCRIPTION
**Pull request by ankitpati**
Changes requested in review of #1265 incorporated.

### CODE REVIEW

File: data/filedefs/filetypes.Swift.conf [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+# multiline comments
+comment_open=/*
+comment_close=*/
+
+comment_use_indent=true
+
+[build-menu]
+FT_00_LB=
+FT_00_CM=
+FT_00_WD=
```
Comment (b4n) [inline data/filedefs/filetypes.Swift.conf]: you should remove the empty entry and move the next ones in place (e.h. `FT_01` → `FT_00` and `FT_02` → `FT_01`)

File: data/Makefile.am [hunk 2]
```diff
@@ -61,6 +61,7 @@ filetypes = \
 	filedefs/filetypes.ruby \
 	filedefs/filetypes.rust \
 	filedefs/filetypes.Scala.conf \
+	filedefs/filetypes.Swift.conf \
```
Comment (b4n) [inline data/Makefile.am]: not very important, but we sort them alphabetically and case insensitively, so it should go after `filetypes.sql`

**Comment by ankitpati** [comment]
Incorporated changes requested in the review.

**Comment by ankitpati** [comment]
Thank you for the approval. Should I squash the commits now?

**Comment by b4n** [comment]
You can, or you can let us do it when merging: either way is fine.

**Comment by ankitpati** [comment]
I am fine with either. What is the norm here?

**Comment by b4n** [comment]
We generally like to ask PR authors to do it for two reasons, one good, and one bad:
* The good reason is that, if there are many commits and not everything should be squashed together but only some fixes to specific commits, the author is more likely to easily know what fix should got to what commit.
* The bad reason is that GitHub doesn't provide a way to mark PRs as merged if squashing manually (e.g. not using their squash button) and not pushed to the PR branch.  Best we can do is close them.

However, we are aware that not everybody is comfortable enough with Git to squash manually without fear or trial and error, so ultimately we are happy to do it for anyone not comfortable doing it him/herself.

Also, last thing: we always prefer squashing to happen at the very end of a PR review process, when everything is sorted out.  Squashing commits during review iterations makes it harder to see changes for reviewers, so apart for really simple commits it's better to wait for approval.  You've got mine for this one so it's fine though :)

So, **the summary is**: if you have no problem doing it, please, do.  Otherwise, don't worry for a second and we'll do it ourselves.

**Comment by elextr** [comment]
Did you try the C++ lexer as I suggested on your previous PR?  I don't know swift, but as I noted the support in Scite (which is from the author of Scintilla) seems to use c++.

**Comment by b4n** [comment]
> Did you try the C++ lexer as I suggested on your previous PR? I don't know swift, but as I noted the support in Scite (which is from the author of Scintilla) seems to use c++.

There's only really one lexer for C and C++ in Scintilla: `SCLEX_CPP`.

**Comment by ankitpati** [comment]
Squashed and force-pushed.

**Comment by ankitpati** [comment]
@b4n May I seek your attention for this PR?

There is a trivial merge conflict. Should I fix it?

**Comment by elextr** [comment]
@ankitpati though the conflict is trivial, might as well fix it, the less for committers to do the more likely it will happen.

LGBI

**Comment by ankitpati** [comment]
Resolved. Please consider merging this in at the earliest.

**Comment by b4n** [comment]
LGBI.

@codebrainz raised concerns on IRC about having "imperfect" filetype (no symbols and related features like autocompletion and calltips) in core and that people would complain it's broken.  I don't really know how used Swift is, and I don't really think it's bets not to have support for a language than having partial one.  Opinions? @elextr 

BTW, I started a CTags parser for this, but I won't likely have it ready too quickly.  But @codebrainz suggested maybe someone could quickly write a regex-based parser based off some ctags command line.  Should be fairly easy I think.

**Comment by elextr** [comment]
Well, this PR will give at least "programmers editor" level of support, even if it has no symbols, so it is a useful thing.

I don't know swift, but IIUC its a derivative from obj-c++ so I would wonder how well a regex based parser would work.  Such a parser would not work on C/C++ of course.

**Comment by ankitpati** [comment]
Thank you. Looking into it.

**Comment by codebrainz** [comment]
@ankitpati, these are some useful docs for adding parsers from Exuberant CTags project (which Universal CTags is a fork of), I presume they mostly still apply:

http://ctags.sourceforge.net/parser.html
http://ctags.sourceforge.net/EXTENDING.html

There's this VIM config file which shows the command-line based custom Swift regex parser some people use (no idea how well it works), which could be ported to a compiled regexp parser as per above docs:

https://github.com/keith/swift.vim/commit/5d5f81483d1ecc611c0dcd2846aab872f02ab074

According to @b4n the main limitation with a regexp-based parser is that it can't handle stuff like scope information, while a character-based parser can support this since it can maintain own state during the parse, if I understood correctly.

**Comment by ankitpati** [comment]
@codebrainz It is possible to maintain state with regex using capturing groups and look-arounds, although the feasibility of implementing a complex parser is debatable.

**Comment by elextr** [comment]
@ankitpati note that @b4n mentioned above starting a parser, maybe you should lean on him to put his "work in progress" on github so you can assist with that.  Better than two people working on it.

**Comment by ankitpati** [comment]
@elextr Agreed. @b4n Any thoughts?

**Comment by b4n** [comment]
Yeah sure.  I pushed the current state I have to https://github.com/b4n/fishman-ctags/tree/wip/parsers/swift (it's against [Universal-CTags](https://github.com/universal-ctags/ctags)).  I'm not sure what's the exact state, but it surely is full of holes.
Also beware, it uses features not yet imported in Geany, so integrating it will require merging #1263 first.

Useful stuff I used:
* More or less strict language reference: https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/AboutTheLanguageReference.html#//apple_ref/doc/uid/TP40014097-CH29-ID345
* test stuff when you don't have a clue how to install swift: https://swiftlang.ng.bluemix.net/#/repl

**Comment by ankitpati** [comment]
> test stuff when you don't have a clue how to install swift

I could have [helped](https://swift.org/download/#using-downloads "Swift.org") with that.

After that, you may actually use [`filetype_extensions.conf`](https://raw.githubusercontent.com/ankitpati/geany/2a86cad6f3219e02a967a357dafabae5c553bfdb/data/filetype_extensions.conf "PR 1323") and [`filetypes.Swift.conf`](https://raw.githubusercontent.com/ankitpati/geany/2a86cad6f3219e02a967a357dafabae5c553bfdb/data/filedefs/filetypes.Swift.conf "PR 1323") from this very PR to get Swift support in Geany.

**Comment by ntrel** [comment]
> I guess it's better to have partial support for a language than no support at all. Opinions? 

Absolutely, highlighting without tag parsing is fine. And once highlighting is in, people will start to use Swift with Geany, so someone else may step up to add tag parsing.

File: data/filedefs/filetypes.Swift.conf [hunk 3]
```diff
@@ -0,0 +1,42 @@
+# For complete documentation of this file, please see Geany's main documentation
+[styling=C]
+
+[keywords]
+primary=associativity break case catch class continue convenience default deinit didSet do else enum extension fallthrough false final for func get guard if import in infix init inout internal lazy let mutating nil operator override postfix precedence prefix private public repeat required return self set static struct subscript super switch throws true try var weak where while willSet
+secondary=Array Bool Dictionary ErrorType Int Float Double Set String Tuple UnicodeScalar abs max min print
+# documentation keywords for javadoc
+doccomment=author deprecated exception param return see serial serialData serialField since throws todo version
+
```
Comment (ntrel) [inline data/filedefs/filetypes.Swift.conf]: Could also add this for `"""multiline string"""` highlighting:
```ini
[lexer_properties]
lexer.cpp.triplequoted.strings=1
```

Comment (ankitpati) [reply data/filedefs/filetypes.Swift.conf]: Done.

**Comment by ntrel** [comment]
Thanks @ankitpati.

@b4n et al: OK to merge? (got thumbs up from @codebrainz on my previous comment).

**Comment by elextr** [comment]
LGBI

**Pull request merged**